src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -1531,9 +1543,55 @@ Collection<CompactionAggregate.UnifiedAggregate> getCompactionAggregates(Arena a
                                                                       this::makeBucket,
                                                                       unbucketed::addAll);
 
-            List<CompactionAggregate.UnifiedAggregate> aggregates = new ArrayList<>();
-            for (Bucket bucket : buckets)
-                aggregates.add(bucket.constructAggregate(controller, spaceAvailable, arena));
+            if (!buckets.isEmpty())
+            {
+                for (Bucket bucket : buckets)
+                    aggregates.add(bucket.constructAggregate(controller, spaceAvailable, arena));
+            }
+            else
+            {
+                // If there are no overlaps, we look if some shards have too many SSTables.
+                // If that's the case, we perform a major compaction on those shards.
+                List<Set<CompactionSSTable>> groups =
+                shardManager.splitSSTablesInShards(sstables,
+                                                   new ShardingStats(sstables, shardManager, controller).shardCountForDensity,
+                                                   (sstableShard, shardRange) -> Sets.newHashSet(sstableShard));
+
+                List<Set<CompactionSSTable>> oversizeGroups =
+                groups.stream()
+                      .filter(group -> group.size() > threshold * controller.getShardMaxSstablesFactor())
+                      .collect(Collectors.toList());
+
+                if (!oversizeGroups.isEmpty())
+                {
+                    // Now combine the groups that share an sstable so that we have valid independent transactions.
+                    // Only keep the groups that were combined with an oversize group.
+                    groups = Overlaps.combineSetsWithCommonElement(groups);
+                    unbucketed.clear();
+
+                    for (Set<CompactionSSTable> group : groups)
+                    {
+                        if (group.stream().anyMatch(
+                        sstable -> oversizeGroups.stream().anyMatch(oversizeGroup -> oversizeGroup.contains(sstable))))
```

Comment:
AFAIK we can't use streams in OSS C*, where this will also need to be ported. Perhaps turn these into loops already here (IntelliJ can do this for you) so that the code does not immediately diverge?

Reply:
Is this a performance concern ?
Because the min JDK seems to be 11 even for OSS C*

Reply:
Performance. Apparently it is not yet officially forbidden, but it will be: https://lists.apache.org/thread/65glsjzkmpktzmns6j9wvr4nczvskx36

I'm also afraid this can fall under the current "don't use streams on hot paths" recommendation.